File: Android/app/src/main/java/com/finora/app/data/model/RecurringRule.kt
package com.finora.app.data.model

import com.google.gson.annotations.SerializedName

data class RecurringRule(
    @SerializedName("_id") val id: String = "",
    val user: String = "",
    val name: String = "",
    val type: String = "Expense", // Income, Expense, Transfer
    val amount: Double = 0.0,
    val account: Any? = null,
    val toAccount: Any? = null,
    val category: Any? = null,
    val subcategory: Any? = null,
    val merchant: String = "",
    val notes: String = "",
    val frequency: String = "monthly", // daily, weekly, monthly, yearly
    val nextRunDate: String = "",
    val endDate: String? = null,
    val isActive: Boolean = true,
    val autoPost: Boolean = false,
    val createdAt: String? = null,
    val updatedAt: String? = null
) {
    val categoryName: String
        get() = when (category) {
            is Map<*, *> -> (category["name"] as? String) ?: "General"
            else -> "General"
        }

    val categoryColor: String
        get() = when (category) {
            is Map<*, *> -> (category["color"] as? String) ?: "#4F46E5"
            else -> "#4F46E5"
        }

    val accountName: String
        get() = when (account) {
            is Map<*, *> -> (account["name"] as? String) ?: "Account"
            else -> "Account"
        }
}

File: Server/models/RecurringRule.js
import mongoose from 'mongoose';

const recurringRuleSchema = new mongoose.Schema({
  user: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true,
  },
  name: {
    type: String,
    required: true,
    trim: true,
  },
  type: {
    type: String,
    enum: ['Income', 'Expense', 'Transfer'],
    required: true,
  },
  amount: {
    type: Number,
    required: true,
    min: 0.01,
  },
  account: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Account',
    required: true,
  },
  toAccount: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Account',
    default: null,
  },
  category: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Category',
    default: null,
  },
  subcategory: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Category',
    default: null,
  },
  merchant: {
    type: String,
    default: '',
  },
  notes: {
    type: String,
    default: '',
  },
  frequency: {
    type: String,
    enum: ['daily', 'weekly', 'monthly', 'yearly'],
    required: true,
  },
  nextRunDate: {
    type: Date,
    required: true,
  },
  // endDate: optional — after this date, rule becomes inactive
  endDate: {
    type: Date,
    default: null,
  },
  isActive: {
    type: Boolean,
    default: true,
  },
  autoPost: {
    // If true, auto-creates transaction on due date via cron
    // If false, user must manually confirm / "Mark as Paid"
    type: Boolean,
    default: false,
  },
}, {
  timestamps: true,
});

recurringRuleSchema.index({ user: 1, isActive: 1, nextRunDate: 1 });

const RecurringRule = mongoose.model('RecurringRule', recurringRuleSchema);
export default RecurringRule;

